### PR TITLE
Fixed tab switching & "add layer" bugs

### PR DESCRIPTION
**Pull request by ZachTRice**
- [x] Change palettesLegend import to reference a function
- [x] Import googleAnalytics from worldview-components to fix tab error
- [x] Add previously undefined variables errors when adding layer.
- [x] Add check for null variable when passed into ui hide function.
- [x] Add previously undefined variables breaking the colorbar on hover
- [x] Fix layersInfo import naming preventing info window from opening
- [x] Fix layersOptions import naming preventing config window from opening
- [x] Fix `$colorBox is not defined` error in palettes/legend.js
- [x] Fix `palette is not defined` in palettes/model.js
- [x] Fix `ReferenceError: oldLegend is not defined` in palettes/model.js
- [x] Fix `cannot call methods on dialog prior to initialization; attempted to call method 'close'` error when attempting to use options button to open options dialog after info dialoag is open; or vice-versa

After latest updates to module-loaders base branch a few more bugs were introduced:
- [x] Fix nouislider dependency; `nouislider-browser` is v5, `nouislider` starts at v7. We were using v6.1.0
- [x] Fix jQuery dialog box styling affected by change to jQuery-ui-1.10.4
- [x] Fix `cache.removeWhere` "o.removeWhere is not a function" error -- relates to jscache package
- [x] Fix issue with opacity slider performing strangely onClick (slow to move around, and keeps moving after click release)

### CODE REVIEW

File: web/js/layers/active.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ import util from '../util/util';
 import wvui from '../ui/ui';
 import layersInfo from './info';
 import layersOptions from './options';
-import palettesLegend from '../palettes/legend';
+import {palettesLegend} from '../palettes/legend';
```
Comment (localjo) [inline web/js/layers/active.js]: Just curious, why is this change being made? Don't both versions mean the same thing?

File: web/js/util/load.js [hunk 2]
```diff
@@ -36,8 +36,7 @@ export default (function () {
           loading -= 1;
           if (loading === 0) {
             clearTimeout(indicatorTimeout);
-            indicatorTimeout = null;
```
Comment (localjo) [inline web/js/util/load.js]: Was this line causing problems? I wonder if it was in there for browser compatibility?

Comment (ZachTRice) [reply web/js/util/load.js]: After testing, the fix is related to the next line in the code. Passing in a null parameter causes issues with the qvui hide function.

File: web/js/layers/active.js (hunk 1 above)
Comment (ZachTRice) [reply web/js/layers/active.js]: Otherwise I get `Uncaught TypeError: (0 , _legend2.default) is not a function`. To be perfectly fair, I am not entirely sure why it is only working this way. There seems to be quite a few gotchas related to import and export naming which I need to review further to understand what the issue is in this particular case.
https://stackoverflow.com/questions/36795819/when-should-i-use-curly-braces-for-es6-import/36796281

File: web/js/util/load.js (hunk 2 above)
Comment (localjo) [reply web/js/util/load.js]: The next line uses `indicatorId` not `indicatorTimeout`

File: web/js/layers/active.js (hunk 1 above)
Comment (localjo) [reply web/js/layers/active.js]: Ah, I understand now. If you're importing something by name, rather than the default, the `{}` are required.

**Comment by localjo** [comment]
This looks good to me. Could you update the wording of the commit message to use imperative mood before merging?

**Comment by localjo** [comment]
This change seems to cause some problems when adding a layer. When I check the box for a layer I see;
`Uncaught TypeError: Cannot read property 'hide' of undefined                 load.js:40`

**Comment by localjo** [comment]
This looks good to me. Could you update the wording of the commit message to use imperative mood before merging?

**Comment by ZachTRice** [comment]
Commit message has been updated

File: web/js/layers/options.js [hunk 3]
```diff
@@ -273,6 +273,7 @@ export function layersOptions(config, models, layer) {
   };
 
   var onRangeUpdate = function () {
+    var current;
```
Comment (localjo) [inline web/js/layers/options.js]: You can add the `var` keyword on line 283 instead, since that's in the middle of the list of vars being declared for this function scope. If you want the variables to be defined first in this function, you can move the `updateRangeLabels()` call below that list.

File: web/js/palettes/legend.js [hunk 4]
```diff
@@ -59,6 +59,7 @@ export function palettesLegend(spec) {
   };
 
   var renderScale = function ($legendPanel, legend, index, layerId) {
+    var $container, $colorbar;
```
Comment (localjo) [inline web/js/palettes/legend.js]: You can add the `var` keyword to the next two lines where these variables are defined, instead of breaking the variable definitions into a separate line like this.

File: web/js/palettes/legend.js [hunk 5]
```diff
@@ -116,10 +117,10 @@ export function palettesLegend(spec) {
   };
 
   var updateClasses = function (legend, index) {
+    var $colorBox;
```
Comment (localjo) [inline web/js/palettes/legend.js]: You can add the `var` keyword two lines down where `$colorBox` is actually defined.

File: web/js/palettes/model.js [hunk 6]
```diff
@@ -27,6 +27,7 @@ export function palettesModel(models, config) {
   };
 
   self.getCustom = function (paletteId) {
+    var palette;
```
Comment (localjo) [inline web/js/palettes/model.js]: You can add the `var` keyword to the next line instead.

File: web/js/palettes/model.js [hunk 7]
```diff
@@ -346,7 +347,7 @@ export function palettesModel(models, config) {
 
     loEach(state.l, function (layerDef) {
       var layerId = layerDef.id;
-      var minValue, maxValue;
+      var values, minValue, maxValue;
```
Comment (localjo) [inline web/js/palettes/model.js]: Why are you defining `values` outside of the scope it was originally used in? It looks like the line `values = util.toArray(attr.value.split(';'));` should have the `var` keyword added to it to keep the variable in the same scope. Moving it out of that scope could change the behavior.

File: web/js/palettes/model.js [hunk 8]
```diff
@@ -479,12 +480,13 @@ export function palettesModel(models, config) {
   };
 
   var updateLookup = function (layerId) {
+    var legend, oldLegend, entries;
```
Comment (localjo) [inline web/js/palettes/model.js]: You're moving these variables outside of their original scope. Why?

File: web/js/palettes/legend.js (hunk 5 above)
Comment (ZachTRice) [reply web/js/palettes/legend.js]: There are two each methods containing `$colorbox`, so this var needs to be defined at the top in this function.

File: web/js/palettes/model.js [hunk 9]
```diff
@@ -479,12 +480,13 @@ export function palettesModel(models, config) {
...
+    var lookup = {};
```
Comment (localjo) [inline web/js/palettes/model.js]: You're moving this declaration before the ["early return"](https://forum.freecodecamp.org/t/the-return-early-pattern/19364) on the next line. There would be no need to declare it if the function returns early.

File: web/js/palettes/legend.js (hunk 5 above)
Comment (localjo) [reply web/js/palettes/legend.js]: It looks like those are two different `$colorBox` variables that just happen to have the same name. Are they supposed to refer to the exact same element? Doesn't making them use the same variable increase the risk that one might overwrite the other?

Comment (localjo) [reply web/js/palettes/legend.js]: That existing lodash each loop is a little strange. It looks like the same loop happens twice;
```
loEach(legend.colors, function (color, classIndex) {
```
That could probably be simplified to only happen once.

File: web/js/palettes/model.js (hunk 7 above)
Comment (ZachTRice) [reply web/js/palettes/model.js]: Otherwise `values` is undefined. Seems to make sense to define it here.

File: web/js/palettes/model.js (hunk 8 above)
Comment (ZachTRice) [reply web/js/palettes/model.js]: Otherwise these values are undefined and when you use these features in WV, the console throws a `ReferenceError: oldLegend is not defined` error.

File: web/js/palettes/legend.js (hunk 5 above)
Comment (ZachTRice) [reply web/js/palettes/legend.js]: Yea, this is weird. It seems like these two can be combined, I'll look at combining these.

File: web/js/palettes/model.js (hunk 7 above)
Comment (localjo) [reply web/js/palettes/model.js]: Why do you need `values` to be defined outside of the scope it was originally defined in? This `var` declaration belongs inside of the `loEach` loop, on line 358 where it's being given a value for the first time, unless you're using it before that.

File: web/js/palettes/model.js (hunk 8 above)
Comment (localjo) [reply web/js/palettes/model.js]: There are no lines referencing these values outside of their original scope. You should add the `var` keyword inside the `loEach` loop, and you can just add it to the lines where those variables are given values. No need to make it a separate line.

**Comment by localjo** [comment]
Keep in mind that everything inside of a lodash.each loop is inside of a callback function, which means [it's in a new scope](http://2ality.com/2011/02/javascript-variable-scoping-and-its.html). We should be careful about changing variable scopes unintentionally because it could have side effects that very difficult to debug during a refactor.

**Comment by ZachTRice** [comment]
Thanks. I have moved variables back into the each loops for now. Layer sidebar info, option and colorbar functions are all working. There is still a bug when hovering the map where the colorbar label doesnot appearing but this code is related to the map (code in map/runningdata.js) so I am going to document that in the main module loaders branch.

File: web/js/palettes/legend.js [hunk 10]
```diff
@@ -118,11 +118,13 @@ export function palettesLegend(spec) {
...
     $panel.empty();
+    var $detailPanel = $('<div></div>');
     loEach(legend.colors, function (color, classIndex) {
-      var $colorBox;
+      var label = legend.tooltips[classIndex];
+      label = (legend.units) ? label + ' ' + legend.units : label;
       var $runningDataPointLabel = $('<span></span>')
         .addClass('wv-running-category-label');
-      $colorBox = $('<span></span>')
+      var $colorBox = $('<span></span>')
```
Comment (localjo) [inline web/js/palettes/legend.js]: It looks like the `$colorBox` in the first loop and the `$colorBox` in the second loop are different HTML elements, so I think they should use two different variables. You could either give the two elements different variable names, which is probably easiest, or refactor a bit more and eliminate the need to store the "color box" in an intermediate variable at all. What do you think?

**Comment by ZachTRice** [comment]
This PR is a complete mess now after merging in the latest changes to `module-loaders` which includes the addition of imported modules. One of the big issues is the inconsistencies with the packages we were using before.

`jquery-ui-1.11.4` was used before, now we are imprting `jquery-ui-1.10.4`. This has affected the styles of the dialog box, not to mention any other js issues that may arise that aren't evident.

`noUiSlider` was using 6.1.10, now I am trying to import `noUiSlider` 7.0.10 because it is the closest version that is semi-working. NPM linking and importing 6.1.10 fixes some issues but then I am left debugging additional (new) issues related to `cache.removeWhere` function which I am trying to track down.

**Comment by localjo** [comment]
It might be best to track down these problems one at a time. Since we don't have testing in place to confirm that our changes aren't breaking things we need to be very careful about reviewing each other's changes and looking for problems. The version mismatches should have been caught before they were merged. But the nature of trying to refactor an app with years of technical debt and no testing is that we're going to have a lot of bugs to track down.

**Comment by ZachTRice** [comment]
The cache issue and slider issues appear to be only used in this area of the app so I am documenting this issues here. If they end up being larger in scope, then I will create a task in the main branch. The version mismatches are just going to be that way since there aren't equivalent npm packages.

**Comment by ZachTRice** [comment]
@localjo could I get you to look over this PR when you get a chance. I would like to get it merged into the base module-loaders branch. It is currently up to speed with the latest changes in that branch.

File: web/css/main.css [hunk 11]
```diff
@@ -10,7 +10,7 @@
 @import url("../../node_modules/jquery.joyride/jquery.joyride.css");
 @import url("../../node_modules/jquery-jcrop/css/jquery.Jcrop.css");
 @import url("../../node_modules/ol/ol.css");
-@import url("../../node_modules/nouislider-browser/jquery.nouislider.css");
+@import url("../../node_modules/nouislider/distribute/jquery.nouislider.min.css");
```
Comment (localjo) [inline web/css/main.css]: If there's a non-minified version of CSS, you should prefer that, since the build script minifies and generates a source map that points to the original file. If we include the minified version, we lose the ability to use the source map.

File: web/js/layers/options.js [hunk 12]
```diff
@@ -147,8 +147,7 @@ export function layersOptions(config, models, layer) {
         }
       })
       .on('slide', function () {
-        models.layers.setOpacity(layer.id, parseFloat($(this)
-          .val()));
+        models.layers.setOpacity(layer.id, $(this).val());
```
Comment (localjo) [inline web/js/layers/options.js]: Why did you remove the `parseFloat` here?

File: web/js/map/ui.js [hunk 13]
```diff
@@ -337,15 +337,15 @@ export function mapui(models, config) {
...
    *
-   * @method updateLayerOrder
+   * @method updateLayer
    * @static
    *
    *
    * @returns {void}
    */
-  var updateLayerOrder = function () {
+  var updateLayer = function () {
```
Comment (localjo) [inline web/js/map/ui.js]: Why did you rename this function? `updateLayerOrder` and `updateLayer` mean different things. Is this function doing something different than it was before?

File: web/js/map/ui.js [hunk 14]
```diff
@@ -80,13 +80,13 @@ export function mapui(models, config) {
...
-      .on('update', updateLayerOrder);
+      .on('update', updateLayer);
     models.date.events.on('select', updateDate);
     models.palettes.events
-      .on('set-custom', updateLookup)
-      .on('clear-custom', updateLookup)
-      .on('range', updateLookup)
-      .on('update', updateLookup);
+      .on('set-custom', updateLayer)
+      .on('clear-custom', updateLayer)
```
Comment (localjo) [inline web/js/map/ui.js]: Why did you change which function is passed as a callback on these four lines?

File: web/js/map/ui.js (hunk 13 above)
Comment (localjo) [reply web/js/map/ui.js]: If this function is just calling `reloadLayers`, why not use that directly, instead of having it wrapped inside another function?

File: web/js/palettes/legend.js [hunk 15]
```diff
@@ -136,14 +135,15 @@ export function palettesLegend(spec) {
...
       $colorBox.on('mouseout', hideUnitsOnMouseOut);
     });
+    // TODO: Review this each loop. It can probably be removed.
     var $detailPanel = $('<div></div>');
     loEach(legend.colors, function (color, classIndex) {
       var label = legend.tooltips[classIndex];
       label = (legend.units) ? label + ' ' + legend.units : label;
       var $row = $('<div></div>')
-        .addClass('wv-palettes-class-detail')
+        .addClass('$colorBox')
```
Comment (localjo) [inline web/js/palettes/legend.js]: Why are you changing the class on each row to `$colorBox` instead of `wv-palettes-class-detail`?

File: package.json [hunk 16]
```diff
@@ -110,6 +110,7 @@
   },
   "browser": {
     "d3": "./web/ext/d3/d3.js",
+    "jscache": "./web/ext/jscache-gitba01cdc/cache.js",
```
Comment (localjo) [inline package.json]: Can this be removed?

File: web/js/layers/options.js (hunk 12 above)
Comment (ZachTRice) [reply web/js/layers/options.js]: `parseFloat` was affecting the click dragging of the slider. There were performance issues / bugginess. Removing it fixes these issues. The step, min and max are defined so I don't see why it is needed.

File: web/js/map/ui.js (hunk 14 above)
Comment (ZachTRice) [reply web/js/map/ui.js]: When we found out that `cache.removeWhere` was broken, I removed it from `updateLookup`. This only left a call to `reloadLayers` within `updateLookup`. But there was another function only calling `reloadLayers`, `updateLayerOrder` so I removed `updateLookup` and renamed `updateLayerOrder` to `updateLayer` since that is a more accurate description of what is happening.

File: web/js/map/ui.js (hunk 13 above)
Comment (ZachTRice) [reply web/js/map/ui.js]: I tried using it directly, but it wasn't working.

File: web/js/palettes/legend.js (hunk 15 above)
Comment (ZachTRice) [reply web/js/palettes/legend.js]: This was accidentally introduced. Good catch.

File: web/js/map/ui.js (hunk 13 above)
Comment (ZachTRice) [reply web/js/map/ui.js]: Also, see my comment above about the reason behind renaming this function.

**Comment by ZachTRice** [comment]
The iCheck issue was related to a change in the "patched" version of iCheck we were running before. See:
https://github.com/fronteed/icheck/blob/1.x/icheck.js#L157
is not the same as
https://github.com/nasa-gibs/worldview/commit/ba80f504d313b0d5b87cfe848ad17ee723e2eb49#diff-dc550be1377c6cbd4583cac3ad4af6d5R157
